### PR TITLE
Avoid using currentConstraint as predicate for getLayouts call

### LINKED ISSUE #11134: Coordinator OOM when a query scans many partitions of a Hive table
TupleDomain shows up multiple times in a TableScanNode:

* `.tableLayout.layout.promisedPredicate` (bad; a discrete TupleDomain representing all partitions)
* `.tableLayout.layout.compactEffectivePredicate` (good; compacted [here](https://github.com/prestodb/presto/blob/0.206/presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionManager.java#L167))
* `.currentConstraint` (bad; a discrete TupleDomain representing all partitions)

The `TaskUpdateRequest.fragment` contains the plan of a fragment, which contains the TableScanNode. It is transmitted to each task once.

Due to the **two** bad JSONs in the TableScan node, such a plan would be roughly 8.4MB.

Take a query with 16 scans of a table with 10000 partitions (without any partition filter) in a cluster of 500 workers. It takes up 8.4MB/JSON * 1 JSON/task * 500 workers * 16 stages = 70GB

70GB of JSON is generated in short succession, and handed to Jetty client, which provides little back pressure. This translates into many humongous allocations, and causes significant GC pressure. This could also fill up the heap, leading to OOM.

### `.currentConstraint`

Fix is easy. It is only used during planning, and does not need to be transported to workers. #11133

### `.layout.connectorHandle.promisedPredicate`

This is used in `getOrComputePartitions`, and is need for executing delete on Hive.

The cause has to do with iterative optimization. Let's take a look at [`PickTableLayout.planTableScan`](https://github.com/prestodb/presto/blob/0.206/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java#L190)

* The first time it's called
  * `predicate` represents the user-provided predicate in the query. (In general, it represents the predicate from filters above.)
  * `node.currentConstraint` is `true`
  * The deterministic part of `predicate` is decomposed into a `TupleDomain` and an expression representing the remaining expression.
    * The expression representing the remaining expression is thrown out. (Is this a bug?)
    * The intersection of the `TupleDomain` and `node.currentConstraint` is stored in `simplifiedConstraint`. (I don't get what `simplified` is supposed to mean here.)
  * `HiveMetadata.getLayouts` is invoked with `simplifiedConstraint`. It returns a `TableLayoutResult`. 
  * For the purpose of this discussion, `TableLayoutResult` consists of roughly 3 parts: (1) `layout.predicate` (2) `unenforcedConstraint` (3) `layout.handle.promisedPredicate` (4) rest of `layout.handle`.
    * What `1` and `3` represents is not documented. Based on my understanding. They need to satisfy the following requirements. Take Universe = `{all tuples of partition column values in the table}`,
      * ~~`1 ∧ 2` must be equal to `simplifiedConstraint`~~ (This is incorrect)
      * ~~`1` must be equal to the list of partitions `4` physically represents.~~ (This is incorrect)
      * `3 ∧ 2` must be equal to `simplifiedConstraint`
      * `3` must be equal to the list of partitions `4` physically represents.
    * In the current implementation, `1` is implemented by reversing from `4`. This makes it a discrete TupleDomain. [here](https://github.com/prestodb/presto/blob/0.206/presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java#L1407)
    * In the current implementation, `3` is implemented by decomposing `simplifiedConstraint`. i.e. [here](https://github.com/prestodb/presto/blob/0.206/presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java#L1367) and [here](https://github.com/prestodb/presto/blob/0.206/presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionManager.java#L154)
  * In the new `TableScanNode`, `node.currentConstraint` will be set to intersect of `1` and `simplifiedConstraint`, making it discrete.
    * Side question: why is the intersection with `simplifiedPredicate` necessary?
* In subsequent times it's called
  * `node.currentConstraint` is now discrete
  * This make `simplifiedConstraint` discrete
  * This then makes `3` discrete

I'm proposing to compute `1` by decomposing `simplifiedConstraint`, instead of by reversing from `4`. This will make it equivalent to `3`.

### PR DESCRIPTION
**Pull request by haozhun**
Fixes #11134

For the query mentioned in #11134: a query with 16 scans of a table with 10000 partitions (without any partition filter). The query's planning time reduced by 4X, from 115 seconds to 25 seconds. The query's initial task update request size reduced by 600X, from 4.4MB (half of what was claimed in #11134 due to #11133) to 7KB.

### CODE REVIEW

**Comment by kokosing** [comment]
Would it be possible to merge this after: https://github.com/prestodb/presto/pull/11164?

File: presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java [hunk 1]
```diff
@@ -56,4 +60,28 @@ public boolean hasAllOutputs(TableScanNode node)
 
         return columns.containsAll(nodeColumnHandles);
     }
+
+    public static TupleDomain<ColumnHandle> computeEnforced(TupleDomain<ColumnHandle> tupleDomain, TupleDomain<ColumnHandle> unenforcedTupleDomain)
```
[the remark this answers is not in the document]
Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java]: I'll add tests.

Yes, your interpretation is correct. Is there any confusion here on what the `tupleDomain` argument means? I tried to come up with more explicit names, but then figured that the semantics is sufficiently as is in the context of this method. Maybe I'm wrong. Do you have any recommendations?

File: presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java [hunk 2]
```diff
@@ -56,4 +60,28 @@ public boolean hasAllOutputs(TableScanNode node)
...
+    {
+        if (tupleDomain.isNone()) {
+            checkArgument(unenforcedTupleDomain.isNone() || unenforcedTupleDomain.isAll());
```
[the remark this answers is not in the document]
Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java]: If the engine requests a connector provides a layout with a domain of "none". The connector can have two possible reactions, either:

1. It can provide an empty table layout, in which case there would be no unenforced domain
2. It cannot. In which case, it will tell the engine to help it filter everything out by providing a "false" unenforced domain.

I had this here as a sanity check to assert one of these two scenarios. It would be invalid argument otherwise. I'll put my answer above as comment in the code.

File: presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java [hunk 3]
```diff
@@ -56,4 +60,28 @@ public boolean hasAllOutputs(TableScanNode node)
...
+        }
+        if (unenforcedTupleDomain.isNone()) {
+            return TupleDomain.none();
+        }
+        Map<ColumnHandle, Domain> domains = tupleDomain.getDomains().get();
+        Map<ColumnHandle, Domain> unenforcedDomains = unenforcedTupleDomain.getDomains().get();
+        ImmutableMap.Builder<ColumnHandle, Domain> enforcedDomainsBuilder = ImmutableMap.builder();
+        for (Map.Entry<ColumnHandle, Domain> entry : domains.entrySet()) {
+            ColumnHandle columnHandle = entry.getKey();
+            if (!unenforcedDomains.containsKey(columnHandle)) {
```
[the remark this answers is not in the document]
Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java]: I don't think so. See `TupleDomain.normalizeAndCopy`.

File: presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java [hunk 4]
```diff
@@ -56,4 +60,28 @@ public boolean hasAllOutputs(TableScanNode node)
...
+            checkArgument(unenforcedTupleDomain.isNone() || unenforcedTupleDomain.isAll());
+            return TupleDomain.none();
+        }
+        if (unenforcedTupleDomain.isNone()) {
+            return TupleDomain.none();
+        }
+        Map<ColumnHandle, Domain> domains = tupleDomain.getDomains().get();
+        Map<ColumnHandle, Domain> unenforcedDomains = unenforcedTupleDomain.getDomains().get();
+        ImmutableMap.Builder<ColumnHandle, Domain> enforcedDomainsBuilder = ImmutableMap.builder();
+        for (Map.Entry<ColumnHandle, Domain> entry : domains.entrySet()) {
```
[the remark this answers is not in the document]
Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java]: `predicateDomain` would be `entry.getValue()`, not `entry`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/BeginTableWrite.java [hunk 5]
```diff
@@ -173,23 +175,24 @@ private PlanNode rewriteDeleteTableScan(PlanNode node, TableHandle handle)
         {
             if (node instanceof TableScanNode) {
                 TableScanNode scan = (TableScanNode) node;
+                TupleDomain<ColumnHandle> originalEnforcedConstraint = scan.getEnforcedConstraint();
```
[the remark this answers is not in the document]
Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/BeginTableWrite.java]: I don't understand why you would want to inline this given this variable has a meaningful name and multiple usages.

File: presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java [hunk 6]
```diff
@@ -56,4 +60,28 @@ public boolean hasAllOutputs(TableScanNode node)
...
+            ColumnHandle columnHandle = entry.getKey();
```
[the remark this answers is not in the document]
Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java]: In general, I prefer to assign `entry.getKey` and `entry.getValue` to variables so that they can have a name. But I also acknowledge that in Presto, we prefer strongly toward inlining, as a result, I only do this if a variable is used multiple times, as is the case here.

File: presto-main/src/main/java/com/facebook/presto/server/remotetask/HttpRemoteTask.java [hunk 7]
```diff
@@ -89,9 +89,9 @@
 import static com.google.common.util.concurrent.Futures.immediateFuture;
```
[the remark this answers is not in the document]
Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/server/remotetask/HttpRemoteTask.java]: I included it here because the end goal of this PR is to reduce this number. I had some measurement, but I forgot to put the result in the description of this PR.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java [hunk 8]
```diff
@@ -183,56 +178,30 @@ public Result apply(TableScanNode tableScanNode, Captures captures, Context cont
...
+            Session session,
+            TypeProvider types,
+            PlanNodeIdAllocator idAllocator,
+            Metadata metadata,
+            SqlParser parser,
+            DomainTranslator domainTranslator)
     {
-        Expression deterministicPredicate = filterDeterministicConjuncts(predicate);
-        DomainTranslator.ExtractionResult decomposedPredicate = DomainTranslator.fromPredicate(
+        return AddExchanges.pickTableLayout(
```
[the remark this answers is not in the document]
Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java]: I'll swap it the other way around for now. I can see your concern, but I don't think a separate class would be much better. After all, it's closely related to these two classes, and not something generic.

I chose to put it in `AddExchange` initially because that was the more up-to-date version. However, `AddExchange` is such a behemoth that having `PickTableLayout` in there can indeed be concerning.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java [hunk 9]
```diff
@@ -1345,6 +1268,99 @@ else if (match2.isPresent()) {
         };
     }
 
+    public static List<PlanNode> pickTableLayout(
```
[the remark this answers is not in the document]
Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: I'll rename

File: presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java]: I think `predicate` instead of `tupleDomain` is a better name, because it is a predicate we are trying to push down.

File: presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java]: > If the engine requests a connector provides a layout with a domain of "none"

It should not happen. `com.facebook.presto.metadata.MetadataManager#getLayouts` will return an empty list for domain `none`.

File: presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java]: yes, but `predicateDomain.getValue()` seems logical. Even`predicateDomain.getKey()` looks good in this context IMO.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/BeginTableWrite.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/BeginTableWrite.java]: Because the name is no shorter than the code it wraps. Code to me is better because it is more flexible. Anyway, up to you.

File: presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java]: Done.

**Comment by haozhun** [comment]
@martint ready for review

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java [hunk 10]
```diff
@@ -183,55 +205,157 @@ public Result apply(TableScanNode tableScanNode, Captures captures, Context cont
...
 
-        TupleDomain<ColumnHandle> simplifiedConstraint = decomposedPredicate.getTupleDomain()
+        TupleDomain<ColumnHandle> newDomain = decomposedPredicate.getTupleDomain()
                 .transform(node.getAssignments()::get)
                 .intersect(node.getCurrentConstraint());
 
+        Map<ColumnHandle, Symbol> assignments = ImmutableBiMap.copyOf(node.getAssignments()).inverse();
+
+        Constraint<ColumnHandle> constraint;
+        if (pruneWithPredicateExpression) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java]: We do this because invoking the expression interpreter is currently very expensive (it used to be fast, fwiw). Add a comment explaining that and a TODO to get rid of this branch once we optimize the interpreter to make it cheap enough to call always.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java [hunk 11]
```diff
@@ -167,6 +167,11 @@ private boolean arePlansSame(FilterNode filter, TableScanNode tableScan, PlanNod
             }
 
             TableScanNode rewrittenTableScan = (TableScanNode) rewrittenFilter.getSource();
+
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java]: Typo in commit message: "guranteed"

File: presto-main/src/main/java/com/facebook/presto/server/remotetask/HttpRemoteTask.java [hunk 12]
```diff
@@ -495,22 +495,23 @@ private synchronized void sendUpdate()
 
         List<TaskSource> sources = getSources();
 
-        Optional<PlanFragment> fragment = Optional.empty();
-        if (sendPlan.get()) {
-            fragment = Optional.of(planFragment);
-        }
+        Optional<PlanFragment> fragment = sendPlan.get() ? Optional.of(planFragment) : Optional.empty();
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/server/remotetask/HttpRemoteTask.java]: Unrelated change

File: presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java [hunk 13]
```diff
@@ -14,13 +14,17 @@
 package com.facebook.presto.metadata;
```
[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java]: > More details can be found in issue 11134.

Use # when referencing the issue so github turns it into a link.

> [...] this commit reduces flexibility of unenforcedConstraint
that a connector can return in getTableLayouts by requiring connector.

Sentence is incomplete.

File: presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java [hunk 14]
```diff
@@ -56,4 +60,54 @@ public boolean hasAllOutputs(TableScanNode node)
...
+
+    public static TupleDomain<ColumnHandle> computeEnforced(TupleDomain<ColumnHandle> predicate, TupleDomain<ColumnHandle> unenforced)
+    {
+        if (predicate.isNone()) {
+            // If the engine requests that the connector provides a layout with a domain of "none". The connector can have two possible reactions, either:
+            // 1. The connector can provide an empty table layout.
+            //   * There would be no unenforced predicate, i.e., unenforced predicate is TupleDomain.all().
+            //   * The predicate was successfully enforced. Enforced predicate would be same as predicate: TupleDomain.none().
+            // 2. The connector can't/won't.
+            //   * The connector would tell the engine to put a filter on top of the scan, i.e., unenforced predicate is TupleDomain.non().
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java]: Typo: non()

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/TableScanNode.java [hunk 15]
```diff
@@ -40,23 +39,26 @@
         extends PlanNode
 {
     private final TableHandle table;
-    private final Optional<TableLayoutHandle> tableLayout;
     private final List<Symbol> outputSymbols;
     private final Map<Symbol, ColumnHandle> assignments; // symbol -> column
 
+    private final Optional<TableLayoutHandle> tableLayout;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/TableScanNode.java]: Unrelated change?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/TableScanNode.java [hunk 16]
```diff
@@ -40,23 +39,26 @@
...
+
     // Used during predicate refinement over multiple passes of predicate pushdown
     // TODO: think about how to get rid of this in new planner
     private final TupleDomain<ColumnHandle> currentConstraint;
 
+    private final TupleDomain<ColumnHandle> enforcedConstraint;
+
     @JsonCreator
-    public TableScanNode(
+    public static TableScanNode jsonCreator(
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/TableScanNode.java]: What's a "jsonCreator"? Call this "newInstance", instead.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/TableScanNode.java [hunk 17]
```diff
@@ -74,23 +76,43 @@ public TableScanNode(
...
+
+    private TableScanNode(
+            PlanNodeId id,
+            TableHandle table,
+            List<Symbol> outputs,
+            Map<Symbol, ColumnHandle> assignments,
+            Optional<TableLayoutHandle> tableLayout,
+            TupleDomain<ColumnHandle> currentConstraint,
+            TupleDomain<ColumnHandle> enforcedConstraint,
+            boolean internal)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/TableScanNode.java]: This flag here is a code smell. Looking at the rest of the commit, it seems to be used only in the static factory method above. Move the check there, instead.

**Comment by haozhun** [comment]
Addressed comments

----------------------

When I was reading this code, `TableScanNode` was a pretty confusing class. The constructor of `TableScanNode` had to omit a bunch of sanity checks to make the json creator work, making the invariants not obvious.

I opted to fix this by having a "internal" boolean parameter on the private constructor. A "dummy" Void parameter could achieve the same outcome.

An alternative is to have the json creator not delegate to another constructor, this will solve the above issue. But it carries two problems with it:
1. The json deserializer here should be private. That's not possible due to Jackson constraint. However, Jackson allows using factory method. With this alternative, factory methods doesn't work because factories have to delegate to constructors.
2. Minor code duplication.

I don't feel very strongly here. As a result, I took the above alternative that you suggested.
